File: frontend/csvcheck.py
"""Validate the CSV strings produced by src/csv.ts using Python's trusted parser.

Checks: correct headers, column counts, field round-trip integrity (escaping of
commas/quotes/newlines), scored CSV strips label+attack_type, ground-truth holds
exactly the malicious commands, and the malicious set matches across files.
"""
import csv
import io
import json
import sys

with open("/tmp/csv_check.json") as f:
    cases = json.load(f)

failures = []


def check(cond, msg):
    if not cond:
        failures.append(msg)
    return cond


def parse(text):
    return list(csv.reader(io.StringIO(text)))


for case in cases:
    name = case["name"]
    result = case["result"]
    rows = result["rows"]
    malicious = result["malicious"]

    # --- labeled CSV (4 cols) ---
    lab = parse(case["labeled"])
    check(lab[0] == ["process_name", "command_line", "label", "attack_type"],
          f"[{name}] labeled header wrong: {lab[0]}")
    check(len(lab) - 1 == len(rows), f"[{name}] labeled row count {len(lab)-1} != {len(rows)}")
    for i, (orig, got) in enumerate(zip(rows, lab[1:])):
        check(len(got) == 4, f"[{name}] labeled row {i} has {len(got)} cols")
        check(got == [orig["process_name"], orig["command_line"], orig["label"], orig["attack_type"]],
              f"[{name}] labeled row {i} mismatch:\n  orig={orig}\n  got ={got}")

    # --- scored CSV (2 cols, NO label/attack_type leaking) ---
    sc = parse(case["scored"])
    check(sc[0] == ["process_name", "command_line"], f"[{name}] scored header wrong: {sc[0]}")
    check(len(sc) - 1 == len(rows), f"[{name}] scored row count {len(sc)-1} != {len(rows)}")
    for i, (orig, got) in enumerate(zip(rows, sc[1:])):
        check(len(got) == 2, f"[{name}] scored row {i} has {len(got)} cols (label leaked?)")
        check(got == [orig["process_name"], orig["command_line"]],
              f"[{name}] scored row {i} mismatch")
    # no label/attack_type values anywhere in scored
    flat = [c for row in sc[1:] for c in row]
    check("malicious" not in flat and "benign" not in flat,
          f"[{name}] scored CSV leaks label values")

    # --- ground-truth CSV (3 cols) ---
    gt = parse(case["groundtruth"])
    check(gt[0] == ["process_name", "command_line", "attack_type"],
          f"[{name}] groundtruth header wrong: {gt[0]}")
    check(len(gt) - 1 == len(malicious), f"[{name}] groundtruth count {len(gt)-1} != {len(malicious)}")
    for i, (orig, got) in enumerate(zip(malicious, gt[1:])):
        check(got == [orig["process_name"], orig["command_line"], orig["attack_type"]],
              f"[{name}] groundtruth row {i} mismatch")

    # --- cross-check: every ground-truth command appears as a malicious row ---
    lab_mal = {(r[0], r[1]) for r in lab[1:] if r[2] == "malicious"}
    gt_set = {(r[0], r[1]) for r in gt[1:]}
    check(gt_set <= lab_mal, f"[{name}] ground-truth cmd not found among malicious rows")
    check(len(lab_mal) == sum(1 for r in rows if r["label"] == "malicious"),
          f"[{name}] malicious count mismatch in labeled CSV")

    print(f"[{name}] rows={len(rows)} malicious={len(malicious)} "
          f"-> labeled/scored/groundtruth parsed OK")

print()
if failures:
    print("FAILURES:")
    for f in failures:
        print(" -", f)
    sys.exit(1)
print("ALL CHECKS PASSED ✓")

File: frontend/csvcheck.ts
// Exercises the REAL production CSV builders from src/csv.ts and dumps their
// output to /tmp/csv_check.json for cross-validation with Python's csv parser.
import { toGroundTruthCsv, toLabeledCsv, toScoredCsv } from "./src/csv";
import type { GenerateResult } from "./src/types";

// Adversarial fixture: commas, double-quotes, and a real newline in a field.
const adversarial: GenerateResult = {
  story: "adversarial",
  rows: [
    { process_name: "cmd.exe", command_line: 'echo "a,b" && dir', label: "malicious", attack_type: "test" },
    { process_name: "bash", command_line: "printf 'l1\nl2'", label: "benign", attack_type: "benign" },
    { process_name: "powershell.exe", command_line: 'gci -Include *.docx,*.xlsx', label: "malicious", attack_type: "ransomware" },
    { process_name: "git", command_line: "git commit -m \"fix, again\"", label: "benign", attack_type: "benign" },
  ],
  malicious: [
    { process_name: "cmd.exe", command_line: 'echo "a,b" && dir', attack_type: "test" },
    { process_name: "powershell.exe", command_line: 'gci -Include *.docx,*.xlsx', attack_type: "ransomware" },
  ],
};

async function fetchReal(): Promise<GenerateResult> {
  const url =
    "http://127.0.0.1:8000/api/generate?scenario_ids=ransomware&scenario_ids=data_exfiltration&os=windows";
  const res = await fetch(url);
  if (!res.ok) throw new Error(`backend ${res.status}`);
  return res.json();
}

function pack(name: string, r: GenerateResult) {
  return {
    name,
    result: r,
    labeled: toLabeledCsv(r),
    scored: toScoredCsv(r),
    groundtruth: toGroundTruthCsv(r),
  };
}

const real = await fetchReal();
const out = [pack("adversarial", adversarial), pack("real", real)];

const fs = await import("node:fs");
fs.writeFileSync("/tmp/csv_check.json", JSON.stringify(out, null, 2));
console.log("wrote /tmp/csv_check.json | cases:", out.map((o) => o.name).join(", "));
